### DISCUSSION TITLE
Detecting when the eraser brush has completely erased

### DISCUSSION THREAD

**Discussion by aurpo1** [fabricjs/fabric.js#9674] [Q&A]
Hi!
I'm using the eraser brush, for free drawing sketch app.

I drew an object and erased it using an eraser brush(http://fabricjs.com/erasing). 
I expected the object to completely disappear, but the object still remains. (selected box remains)
![image](https://github.com/fabricjs/fabric.js/assets/80630505/139a5ff7-0670-498c-b04f-093e25c2a8a7)

I want to detect when an object has been completely erased from the eraser brush.
The way I think it was to use 'getContext', 'getImageData' to compare pixel by pixel, but this would take too long.

Is there another way?

Thank you.

**Comment by ShaMan123** [comment]
I don't think so
There are several duplicates of this.
I suggest you run the check in a web worker after you transfer the canvas or image data to it
Then you can send back the answer and remove the object if needed

**Reply by aurpo1** [reply]
I'm going to run my app somewhere other than the web. The performance is poor on my work space.😥  I tested it -> getImageData takes a lot of time.

So What I'm curious about is whether fabric js provides any methods for what I want. (Detect completely erase)
If there is, I would appreciate some guidance. If not, I will find another way.

**Reply by aurpo1** [reply]
Is this the issue you mentioned? https://github.com/fabricjs/fabric.js/discussions/8586

**Reply by ShaMan123** [reply]
Yes and there is another one at least
Which platform?
Please share your findings

**Reply by aurpo1** [reply]
for LG TV webos platform. I guess hardware performance is the reason.
https://github.com/fabricjs/fabric.js/pull/6994#issuecomment-829075360 < Is another one?

**Reply by ShaMan123** [reply]
Yes
You could always do the work in the backend and remove the object async

**Comment by ShaMan123** [comment]
I have this weird idea that might well be a dead end
Try the following algorithm using [bezier-js](https://pomax.github.io/bezierjs/)

1. Define outline to be the object described by paths suited for intersection tests/`bezier-js`
2. Define the eraser path as an array of C paths (Bezier class of `bezier-js`)
3. Loop through the C paths and do the following:

   - test each path against the outline for intersection/being contained
   - in case of intersection, cut the outline using the intersected path as the knife, taking into account stroke width
   - in case of being contained, continue
   - else, continue (will retest on smaller outlines as part of the algorithm if contained in outline)
   - repeat until no more intersections
   - if remained outlines that are not contained, the object is not fully erased

**Reply by aurpo1** [reply]
Thank you for your consideration! I will try it.

**Comment by BalachanderM** [comment]
Did you got any solution for this? even I face the same problem :-(

**Reply by asturur** [reply]
The eraser moved to its own repository. 
https://github.com/ShaMan123/erase2d

**Reply by mhieupham1** [reply]
it is working, thank you

**Reply by BalachanderM** [reply]
is it working for you @mhieupham1 
how to download the code 
and run it ? can you tell me thanks!

**Reply by mhieupham1** [reply]
@BalachanderM In my case, i using fabricjs version 4.6.0 and have override _finalizeAndAddPath function in fabric.EraserBrush 
i leave my sample code here: https://jsfiddle.net/mhieupham/4ztLa296/61/

**Reply by BalachanderM** [reply]
@mhieupham1 Thanks! Will this work in react with typescript

**Reply by mhieupham1** [reply]
@BalachanderM yes. this will work in React or Nextjs with TS.

**Reply by BalachanderM** [reply]
@mhieupham1 I have tried using react with typescript but I couldn't figure it out can you give the codefile if it works for you? thanks, a lot

**Reply by mhieupham1** [reply]
@BalachanderM yes, i have used fabric.js on this repo https://github.com/tbolis/react-sketch, you can use it and add the eraser function, with the eraser code from line 1 to line 760 in https://jsfiddle.net/mhieupham/4ztLa296/61/

**Comment by davidanquetin** [comment]
Hi, 
always looking for a simple way to know if a path is fully erased or not (is there a simple parameter for this ?) when listening to `erasing:end` event ?
I need to keep erased object on canvas (because of undo system).
I tried some different tests (intersections, etc) but any of this is efficient for now.
Has someone find a simple method ?
I use old EraserBrush with Fabric 5.
thanks
